kernel/pipe.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2024 Måns Ansgariusson <mansgariusson@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/internal/syscall_handler.h>
+#include <ksched.h>
+#include <kthread.h>
+#include <wait_q.h>
+
+static inline bool pipe_full(struct k_pipe *pipe)
+{
+	return ring_buf_space_get(&pipe->buf) == 0;
+}
+
+static inline bool pipe_empty(struct k_pipe *pipe)
+{
+	return ring_buf_is_empty(&pipe->buf);
+}
+
+static inline int wait_for_data(_wait_q_t *waitq, struct k_pipe *pipe, k_spinlock_key_t *key,
+				k_timeout_t timeout)
+{
+	if (K_TIMEOUT_EQ(timeout, K_NO_WAIT) || unlikely(pipe->flags & PIPE_FLAG_RESET)) {
+		return -EAGAIN;
+	}
+
+	pipe->waiting++;
+	z_pend_curr(&pipe->lock, *key, waitq, timeout);
+	*key = k_spin_lock(&pipe->lock);
+	pipe->waiting--;
+	if (unlikely(!(pipe->flags & PIPE_FLAG_OPEN))) {
+		return -EPIPE;
+	} else if (unlikely(pipe->flags & PIPE_FLAG_RESET)) {
+		if (!pipe->waiting) {
+			pipe->flags &= ~PIPE_FLAG_RESET;
+		}
+		return -ECANCELED;
+	} else if (!pipe_empty(pipe)) {
+		return 0;
+	}
+	return -EAGAIN;
+}
+
+static inline int wait_for_space(_wait_q_t *waitq, struct k_pipe *pipe, k_spinlock_key_t *key,
+				k_timeout_t timeout)
+{
+	if (K_TIMEOUT_EQ(timeout, K_NO_WAIT) || unlikely(pipe->flags & PIPE_FLAG_RESET)) {
+		return -EAGAIN;
+	}
+
+	pipe->waiting++;
+	z_pend_curr(&pipe->lock, *key, waitq, timeout);
+	*key = k_spin_lock(&pipe->lock);
+	pipe->waiting--;
+	if (unlikely(!(pipe->flags & PIPE_FLAG_OPEN))) {
+		return -EPIPE;
+	} else if (unlikely(pipe->flags & PIPE_FLAG_RESET)) {
+		if (!pipe->waiting) {
+			pipe->flags &= ~PIPE_FLAG_RESET;
+		}
+		return -ECANCELED;
+	} else if (!pipe_full(pipe)) {
+		return 0;
+	}
+	return -EAGAIN;
+}
+
+static void notify_waiter(_wait_q_t *waitq)
+{
+	struct k_thread *thread_to_unblock = z_unpend_first_thread(waitq);
+
+	if (likely(thread_to_unblock)) {
```

Comment:
thread_to_unblock -> thread_to_unblock != NULL

This is from the coding guidelines, Zephyr Rule 85.

> The controlling expression of an if statement and the controlling expression of an iteration-statement shall have essentially Boolean type

Essentially, unless the type being tested is already a boolean, perform an explicit comparison to something of the same type.

This applies elsewhere as well.